File: MultiLanguageSandbox/src/test/deps/php/38.php
<?php

/**
 * Question:
 * You have M identical apples that need to be placed into N identical plates. You are allowed
 * to have some plates empty. How many different ways are there to distribute the apples into the plates?
 * 5,1,1 and 1,5,1 are the same division method.
 *
 * Example:
 * - Input: M = 7, N = 3
 * - Output: 8

 */

 function countDistributionMethods($M, $N) {

    // Create a DP table initialized to 0
    $dp = array_fill(0, $M + 1, array_fill(0, $N + 1, 0));
    
    // Base case: 1 way to distribute 0 apples (all plates empty)
    for ($j = 0; $j <= $N; $j++) {
        $dp[0][$j] = 1;
    }
    
    // Fill the DP table
    for ($i = 1; $i <= $M; $i++) {
        for ($j = 1; $j <= $N; $j++) {
            // Option 1: at least one plate has at least one apple
            // This is equivalent to placing at least one apple in the first plate
            // and then distributing the remaining (i - k) apples into j plates,
            // but to avoid overcounting, we ensure the first plate's count is >= the next plates'
            // So we can think of it as placing 'k' apples in the first plate (k >=1),
            // then the remaining (i -k) apples must be distributed into j plates with each plate <=k.
            // But a better approach is to realize that the problem reduces to partitioning i into parts of size <= min(j, i)
            // So the recurrence is dp[i][j] = dp[i][j-1] (no plates with k apples) + dp[i -k][j] for k from 1 to i/j
            // But to optimize, we can use the following:
            $dp[$i][$j] = $dp[$i][$j - 1]; // plates with 0 apples
            if ($i >= $j) {
                $dp[$i][$j] += $dp[$i - $j][$j];
            }
        }
    }
    
    return $dp[$M][$N];
}
function main(){
    // Test cases
    assert(countDistributionMethods(7, 3) === 8);
    assert(countDistributionMethods(5, 2) === 3);
    assert(countDistributionMethods(10, 5) === 30);

}


main();
?>

File: MultiLanguageSandbox/src/test/deps/java/38.java

public class Solution {

  /**
   * Given an integer n, determine if it is a power of two.
   * If it is, return true; otherwise, return false.
   * A number n is a power of two if there exists an integer x such that n == 2^x.
   *
   * @param n The integer to be checked.
   * @return True if n is a power of two, false otherwise.
   *
   * Example:
   * isPowerOfTwo(1)
   * Output: true
   *
   * Example:
   * isPowerOfTwo(16)
   * Output: true
   *
   * Example:
   * isPowerOfTwo(3)
   * Output: false
   *
   * Example:
   * isPowerOfTwo(4)
   * Output: true
   *
   * Example:
   * isPowerOfTwo(5)
   * Output: false
   */
  public static boolean isPowerOfTwo(int n) 
{
    // A number is a power of two if it is greater than 0 and has exactly one bit set in its binary representation.
    // This can be checked using the expression (n & (n - 1)) == 0.
    // For example:
    // 8 (1000) & 7 (0111) = 0
    // 6 (0110) & 5 (0101) = 4 (not 0)
    return n > 0 && (n & (n - 1)) == 0;
}
  public static void main(String[] args) {
    // Test cases
    assert isPowerOfTwo(1024) == true; // 1024 is 2^10
    assert isPowerOfTwo(65536) == true; // 65536 is 2^16
    assert isPowerOfTwo(14) == false; // 14 is not a power of two

    System.out.println("All tests passed");
  }
}